File: backend/app/orbital/conjunctions.py
"""Conjunction screening: closest approaches between tracked objects.

A miniature space-domain-awareness task. Every tracked spacecraft is propagated
over a short horizon and screened pairwise; for each pair we find the time of
closest approach (TCA), the miss distance, and the relative speed. Real catalogs
screen tens of thousands of objects with smart pre-filters; here the catalog is
small, so a direct O(pairs x steps) sweep with a refined minimum is plenty.

Separation is computed in the TEME inertial frame the propagator already returns,
so no Earth-fixed rotation is needed (both objects share the same frame and time).
"""

from __future__ import annotations

from dataclasses import dataclass

from sgp4.api import Satrec

from .passes import TrackedSat, _jd, build_satrec

# A miss closer than this is flagged as a screening alert.
ALERT_KM = 25.0

Vec3 = tuple[float, float, float]


@dataclass(frozen=True)
class Conjunction:
    id: str
    a_id: int
    a_name: str
    b_id: int
    b_name: str
    tca: float  # epoch ms
    miss_km: float
    rel_speed_km_s: float
    alert: bool


def _state(rec: Satrec, ms: float) -> tuple[Vec3, Vec3] | None:
    jd, fr = _jd(ms)
    err, r, v = rec.sgp4(jd, fr)
    if err != 0:
        return None
    return r, v


def _sep_km(a: Vec3, b: Vec3) -> float:
    return ((a[0] - b[0]) ** 2 + (a[1] - b[1]) ** 2 + (a[2] - b[2]) ** 2) ** 0.5


def _refine_tca(rec_a: Satrec, rec_b: Satrec, lo: float, hi: float) -> tuple[float, float]:
    """Ternary search for minimum separation in [lo, hi] (epoch ms)."""
    for _ in range(40):
        if hi - lo <= 200:
            break
        m1 = lo + (hi - lo) / 3
        m2 = hi - (hi - lo) / 3
        sa1, sb1 = _state(rec_a, m1), _state(rec_b, m1)
        sa2, sb2 = _state(rec_a, m2), _state(rec_b, m2)
        if sa1 is None or sb1 is None or sa2 is None or sb2 is None:
            break
        if _sep_km(sa1[0], sb1[0]) < _sep_km(sa2[0], sb2[0]):
            hi = m2
        else:
            lo = m1
    t = (lo + hi) / 2
    sa, sb = _state(rec_a, t), _state(rec_b, t)
    if sa is None or sb is None:
        return t, float("inf")
    return t, _sep_km(sa[0], sb[0])


def screen_conjunctions(
    sats: list[TrackedSat],
    now_ms: float,
    horizon_hours: float = 6.0,
    step_sec: float = 60.0,
    top_k: int = 12,
) -> list[Conjunction]:
    step_ms = step_sec * 1000.0
    steps = int((horizon_hours * 3600.0) / step_sec)

    recs: list[tuple[TrackedSat, Satrec]] = []
    for s in sats:
        rec = build_satrec(s)
        if rec is not None:
            recs.append((s, rec))

    # Precompute each object's position track once.
    tracks: list[list[Vec3 | None]] = []
    for _s, rec in recs:
        pts: list[tuple[float, float, float] | None] = []
        for i in range(steps + 1):
            st = _state(rec, now_ms + i * step_ms)
            pts.append(st[0] if st is not None else None)
        tracks.append(pts)

    out: list[Conjunction] = []
    for i in range(len(recs)):
        for j in range(i + 1, len(recs)):
            best_idx = -1
            best_d = float("inf")
            for k in range(steps + 1):
                pa, pb = tracks[i][k], tracks[j][k]
                if pa is None or pb is None:
                    continue
                d = _sep_km(pa, pb)
                if d < best_d:
                    best_d, best_idx = d, k
            if best_idx < 0:
                continue
            lo = now_ms + max(0, best_idx - 1) * step_ms
            hi = now_ms + min(steps, best_idx + 1) * step_ms
            tca, miss = _refine_tca(recs[i][1], recs[j][1], lo, hi)

            sa, sb = _state(recs[i][1], tca), _state(recs[j][1], tca)
            rel = 0.0
            if sa is not None and sb is not None:
                rel = (
                    (sa[1][0] - sb[1][0]) ** 2
                    + (sa[1][1] - sb[1][1]) ** 2
                    + (sa[1][2] - sb[1][2]) ** 2
                ) ** 0.5

            a_sat, b_sat = recs[i][0], recs[j][0]
            out.append(
                Conjunction(
                    id=f"cdm-{a_sat.id}-{b_sat.id}-{round(tca)}",
                    a_id=a_sat.id,
                    a_name=a_sat.name,
                    b_id=b_sat.id,
                    b_name=b_sat.name,
                    tca=tca,
                    miss_km=miss,
                    rel_speed_km_s=rel,
                    alert=miss < ALERT_KM,
                )
            )

    out.sort(key=lambda c: c.miss_km)
    return out[:top_k]

File: src/services/conjunctions.ts
// Conjunction screening (frontend engine): closest approaches between tracked
// objects. Mirrors the backend's app/orbital/conjunctions.py so the standalone
// app shows the same space-domain-awareness picture without a server.
//
// Separation is computed in the ECI (TEME) frame satellite.js returns, which is
// frame-consistent for two objects sampled at the same instant.
import * as satellite from 'satellite.js'
import type { Conjunction, TrackedSat } from '../types'

/** A miss closer than this is flagged as a screening alert. */
export const ALERT_KM = 25

type Vec3 = [number, number, number]

function satrecOf(sat: TrackedSat): satellite.SatRec | null {
  try {
    const rec = satellite.twoline2satrec(sat.line1, sat.line2)
    return rec && !rec.error ? rec : null
  } catch {
    return null
  }
}

function stateAt(rec: satellite.SatRec, ms: number): { p: Vec3; v: Vec3 } | null {
  const pv = satellite.propagate(rec, new Date(ms))
  if (!pv || typeof pv.position === 'boolean' || typeof pv.velocity === 'boolean') return null
  if (!pv.position || !pv.velocity) return null
  return {
    p: [pv.position.x, pv.position.y, pv.position.z],
    v: [pv.velocity.x, pv.velocity.y, pv.velocity.z],
  }
}

function sep(a: Vec3, b: Vec3): number {
  return Math.hypot(a[0] - b[0], a[1] - b[1], a[2] - b[2])
}

// Ternary search for the minimum separation time inside [lo, hi] (epoch ms).
function refineTca(a: satellite.SatRec, b: satellite.SatRec, lo: number, hi: number): number {
  for (let k = 0; k < 40 && hi - lo > 200; k++) {
    const m1 = lo + (hi - lo) / 3
    const m2 = hi - (hi - lo) / 3
    const sa1 = stateAt(a, m1)
    const sb1 = stateAt(b, m1)
    const sa2 = stateAt(a, m2)
    const sb2 = stateAt(b, m2)
    if (!sa1 || !sb1 || !sa2 || !sb2) break
    if (sep(sa1.p, sb1.p) < sep(sa2.p, sb2.p)) hi = m2
    else lo = m1
  }
  return (lo + hi) / 2
}

export function screenConjunctions(
  sats: TrackedSat[],
  nowMs: number,
  opts: { horizonHours?: number; stepSec?: number; topK?: number } = {},
): Conjunction[] {
  const horizonHours = opts.horizonHours ?? 6
  const stepSec = opts.stepSec ?? 60
  const topK = opts.topK ?? 12
  const stepMs = stepSec * 1000
  const steps = Math.floor((horizonHours * 3600) / stepSec)

  const recs = sats
    .map((s) => ({ sat: s, rec: satrecOf(s) }))
    .filter((r): r is { sat: TrackedSat; rec: satellite.SatRec } => r.rec !== null)

  // Precompute each object's position track once.
  const tracks: (Vec3 | null)[][] = recs.map(({ rec }) => {
    const pts: (Vec3 | null)[] = []
    for (let i = 0; i <= steps; i++) {
      const st = stateAt(rec, nowMs + i * stepMs)
      pts.push(st ? st.p : null)
    }
    return pts
  })

  const out: Conjunction[] = []
  for (let i = 0; i < recs.length; i++) {
    for (let j = i + 1; j < recs.length; j++) {
      let bestIdx = -1
      let bestD = Infinity
      for (let k = 0; k <= steps; k++) {
        const pa = tracks[i][k]
        const pb = tracks[j][k]
        if (!pa || !pb) continue
        const d = sep(pa, pb)
        if (d < bestD) {
          bestD = d
          bestIdx = k
        }
      }
      if (bestIdx < 0) continue
      const lo = nowMs + Math.max(0, bestIdx - 1) * stepMs
      const hi = nowMs + Math.min(steps, bestIdx + 1) * stepMs
      const tca = refineTca(recs[i].rec, recs[j].rec, lo, hi)

      const sa = stateAt(recs[i].rec, tca)
      const sb = stateAt(recs[j].rec, tca)
      if (!sa || !sb) continue
      const missKm = sep(sa.p, sb.p)
      const relSpeedKmS = Math.hypot(sa.v[0] - sb.v[0], sa.v[1] - sb.v[1], sa.v[2] - sb.v[2])

      const a = recs[i].sat
      const b = recs[j].sat
      out.push({
        id: `cdm-${a.id}-${b.id}-${Math.round(tca)}`,
        aId: a.id,
        aName: a.name,
        bId: b.id,
        bName: b.name,
        tca,
        missKm,
        relSpeedKmS,
        alert: missKm < ALERT_KM,
      })
    }
  }

  out.sort((x, y) => x.missKm - y.missKm)
  return out.slice(0, topK)
}
